plugins/checksum/checksum.c

```diff
@@ -22,216 +22,164 @@
+	g_free (sha256);
+	g_clear_error (&error);
 }
 
-static gboolean
-sha256_from_stream (GFileInputStream *file_stream, char out_buf[])
+static void
+thread_sha256_file (GTask *task, GFile *file, gpointer task_data, GCancellable *cancellable)
 {
 	GChecksum *checksum;
-	gssize bytes_read;
-	guint8 digest[SHA256_DIGEST_LENGTH];
-	gsize digest_len = sizeof(digest);
-	guchar buffer[BUFSIZE];
-	gsize i;
+	GFileInputStream *istream;
+	guint8 buffer[32768];
+	GError *error = NULL;
+	gssize ret;
+
+	istream = g_file_read (file, cancellable, &error);
+	if (error) {
+		g_task_return_error (task, error);
+		return;
+	}
 
 	checksum = g_checksum_new (G_CHECKSUM_SHA256);
 
-	while ((bytes_read = g_input_stream_read (G_INPUT_STREAM (file_stream), buffer, sizeof (buffer), NULL, NULL)))
-	{
-		if (bytes_read == -1)
-		{
-			g_checksum_free (checksum);
-			return FALSE;
-		}
+	while ((ret = g_input_stream_read (G_INPUT_STREAM (istream), buffer, sizeof(buffer), cancellable, &error)) > 0)
+		g_checksum_update (checksum, buffer, ret);
 
-		g_checksum_update (checksum, buffer, bytes_read);
+	if (error) {
+		g_checksum_free (checksum);
+		g_task_return_error (task, error);
+		return;
 	}
 
-	g_checksum_get_digest (checksum, digest, &digest_len);
+	g_task_return_pointer (task, g_strdup (g_checksum_get_string (checksum)), g_free);
 	g_checksum_free (checksum);
-
-	for (i = 0; i < SHA256_DIGEST_LENGTH; i++)
-	{
-		/* out_buf will be exactly SHA256_BUFFER_LENGTH including null */
-		g_sprintf (out_buf + (i * 2), "%02x", digest[i]);
-	}
-
-	return TRUE;
 }
 
-static gboolean
-sha256_from_file (char *filename, char out_buf[])
+static int
+dccrecv_cb (char *word[], void *userdata)
 {
-	GFileInputStream *file_stream;
+	GTask *task;
 	char *filename_fs;
 	GFile *file;
+	hexchat_context *ctx;
+	const char *dcc_completed_dir;
+	char *filename;
+
+	if (hexchat_get_prefs (ph, "dcc_completed_dir", &dcc_completed_dir, NULL) == 1 && dcc_completed_dir[0] != '\0')
+		filename = g_build_filename (dcc_completed_dir, word[1], NULL);
+	else
+		filename = g_strdup (word[2]);
 
 	filename_fs = g_filename_from_utf8 (filename, -1, NULL, NULL, NULL);
-	if (!filename_fs)
-	{
+	if (!filename_fs) {
 		hexchat_printf (ph, "Checksum: Invalid filename (%s)\n", filename);
-		return FALSE;
-	}
-
-	file = g_file_new_for_path (filename_fs);
-	g_free (filename_fs);
-	if (!file)
-	{
-		hexchat_printf (ph, "Checksum: Failed to open %s\n", filename);
-		return FALSE;
+		g_free (filename);
+		return HEXCHAT_EAT_NONE;
 	}
 
-	if (!check_limit (file))
-	{
-		hexchat_printf (ph, "Checksum: %s is larger than size limit. You can increase it with /CHECKSUM SET.\n", filename);
-		g_object_unref (file);
-		return FALSE;
+	/* Print in the privmsg tab of the sender */
+	ctx = hexchat_find_context (ph, NULL, word[3]);
+	// if we didn't request the file from the channel, it's send to us from, use the current one, from which we actually requested it
+	if(!ctx){
+		ctx = hexchat_find_context (ph, NULL, NULL);
 	}
```

Comment:
Maybe we should just store the network id and username. Then later find the correct context.

This is a bit more complicated but you can copy this function here:
https://github.com/hexchat/hexchat/blob/847e5a14d6d7991682c3add853f7b3cc32f935df/plugins/fishlim/plugin_hexchat.c#L83-L107

Reply:
Hmm, well that function doesn't default to the front context on the network if unfound. So will need some tweaking.

Reply:
For simplicity sake you could just store the server name and username and then use `hexchat_find_context()`. Falling back to `hexchat_find_context (ph, servername, NULL);` if needed.

Reply:
Seems reasonable, I already thought about doing that, but what should it do, when it also can't find the servername, should it default to the current channel, which I thought is fine, but you said it's wrong, so should it print nothing, which seems also reasonable :thinking:

Reply:
I now did it the simple way. It also solves the issue of a potentially sending a PRIVMSG to the wrong place.

It now stores the servername and the channel and uses these later to find a context. The channel and servername can be potentially NULL, if the initial `hexchat_get_info` returns NULL, as far as I understand this isn't possible, if we just got a dccoffer, but if we send one? I just wanted to ask, if thats possible, and how to handle those cases, I would say to return and print nothing when the start channel and/or servername are NULL to begin with.